contracts/src/v0.8/keystone/KeystoneForwarder.sol

```diff
@@ -152,27 +146,33 @@ contract KeystoneForwarder is IForwarder, ConfirmedOwner, TypeAndVersionInterfac
     }
 
     bytes32 workflowExecutionId;
-    bytes2 reportId;
     bytes32 configId;
-    uint8 f;
+    bytes32 combinedId;
     {
       uint32 donId;
       uint32 configVersion;
+      bytes2 reportId;
       (workflowExecutionId, donId, configVersion, reportId) = _getMetadata(
         rawReport
       );
 
       configId = keccak256(abi.encode(donId, configVersion));
 
-      f = s_configs[configId].f;
+      uint8 f = s_configs[configId].f;
       // f can never be 0, so this means the config doesn't actually exist
       if (f == 0) revert InvalidDonId(donId);
-    }
 
-    bytes32 combinedId = _combinedId(receiverAddress, workflowExecutionId, reportId);
-    if (s_reports[combinedId].transmitter != address(0)) revert AlreadyProcessed(combinedId);
+      if (f + 1 != signatures.length) revert InvalidSignatureCount(f + 1, signatures.length);
 
-    if (f + 1 != signatures.length) revert InvalidSignatureCount(f + 1, signatures.length);
+      combinedId = _combinedId(receiverAddress, workflowExecutionId, reportId);
+    }
+
+    // TODO: make router address a compile time thing
+    IRouter router = IRouter(address(0));
+    {
+      bool delivered = router.getDeliveryStatus(combinedId);
+      if (delivered) revert AlreadyProcessed(combinedId);
+    }
```

Comment:
You could check this on router (reduces vars)

Reply:
I was considering that, but it would happen after the signature validation so it would cost more gas in the failure case. Maybe it's still worth it?